### PR TITLE
chore: optimize message processor and tally

### PR DESCRIPTION
**Pull request by 0xmad**
# Description

- [x] Remove sha256 hashing
- [x] Support multiple params for verifier

## Additional Notes

N/A

## Related issue(s)

Related to #1614 

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: circuits/circom/core/non-qv/processMessages.circom [hunk 1]
```diff
@@ -48,17 +47,10 @@ include "../../trees/incrementalQuinaryTree.circom";
...
-
-    // The SHA-256 hash of values provided by the contract.
-    signal input inputHash;
-    signal input packedVals;
     // Number of users that have completed the sign up.
-    signal numSignUps;
+    signal input numSignUps;
     // Number of options for this poll.
-    signal maxVoteOptions;
+    signal input maxVoteOptions;
```
Comment (ctrlc03) [inline circuits/circom/core/non-qv/processMessages.circom:53]: given maxVoteOptions is always 5 ** voteOptionTreeDepth we can probably get rid of it completely as a input and just use the calculated one inside the circuit (MESSAGE_TREE_ARITY ** voteOptionTreeDepth)

File: contracts/contracts/MessageProcessor.sol [hunk 2]
```diff
@@ -119,18 +119,9 @@ contract MessageProcessor is Ownable, SnarkCommon, Hasher, CommonUtilities, IMes
...
-        messageRoot,
-        sbCommitment,
-        _newSbCommitment,
-        messageTreeSubDepth,
-        messageTreeDepth,
-        voteOptionTreeDepth,
-        _proof
-      )
-    ) {
+    uint256[] memory publicCircuitInputs = getPublicCircuitInputs(currentMessageBatchIndex, _newSbCommitment);
```
Comment (ctrlc03) [inline contracts/contracts/MessageProcessor.sol:122]: you should be able to pass uint8 messageTreeSubDepth, uint8 messageTreeDepth so we don't need to fetch twice, same for these values:  batchEndIndex, _currentMessageBatchIndex

Comment (ctrlc03) [reply contracts/contracts/MessageProcessor.sol:122]: same for messageRoot

File: contracts/contracts/Tally.sol [hunk 3]
```diff
@@ -154,59 +116,60 @@ contract Tally is Ownable, SnarkCommon, CommonUtilities, Hasher, DomainObjs {
...
-
     (uint256 numSignUps, ) = poll.numSignUpsAndMessages();
 
     // Require that there are untallied ballots left
     if (batchStartIndex >= numSignUps) {
       revert AllBallotsTallied();
     }
 
-    bool isValid = verifyTallyProof(_proof, numSignUps, batchStartIndex, tallyBatchSize, _newTallyCommitment);
+    uint256[] memory circuitPublicInputs = getPublicCircuitInputs(_newTallyCommitment);
```
Comment (ctrlc03) [inline contracts/contracts/Tally.sol:126]: you already have numSignups, batchStartIndex, tallyBatchSize in this scope, can you pass the directly to verifyTallyProof?

File: contracts/contracts/Tally.sol [hunk 4]
```diff
@@ -154,59 +116,60 @@ contract Tally is Ownable, SnarkCommon, CommonUtilities, Hasher, DomainObjs {
...
-  /// @param _tallyBatchSize batch size for the tally
-  /// @param _newTallyCommitment the tally commitment to be verified at a given batch index
   /// @return isValid whether the proof is valid
   function verifyTallyProof(
-    uint256[8] calldata _proof,
-    uint256 _numSignUps,
-    uint256 _batchStartIndex,
-    uint256 _tallyBatchSize,
-    uint256 _newTallyCommitment
+    uint256[] memory _publicInputs,
```
Comment (ctrlc03) [inline contracts/contracts/Tally.sol:163]: given we know how many inputs, maybe we constraint the array to a fixed size?

File: contracts/contracts/crypto/Verifier.sol [hunk 5]
```diff
@@ -26,15 +26,15 @@ contract Verifier is IVerifier, SnarkConstants, SnarkCommon {
...
+  /// @param inputs The public inputs to the circuit
   /// @return isValid Whether the proof is valid given the verifying key and public
   ///          input. Note that this function only supports one public input.
   ///          Refer to the Semaphore source code for a verifier that supports
   ///          multiple public inputs.
   function verify(
     uint256[8] memory _proof,
     VerifyingKey memory vk,
-    uint256 input
+    uint256[] memory inputs
```
Comment (ctrlc03) [inline contracts/contracts/crypto/Verifier.sol:37]: would calldata work here?

**Review by ctrlc03 — COMMENTED** [review]
e2e tests work locally 🤔

File: contracts/contracts/MessageProcessor.sol (hunk 2 above)
Comment (0xmad) [reply contracts/contracts/MessageProcessor.sol:122]: The problem here is related to stack issue I mentioned before + `getPublicCircuitInputs` should be called without additional data passed as params. This is a public method which is also using for proof verification and public inputs check for prover and cli functions.

File: contracts/contracts/Tally.sol (hunk 3 above)
Comment (0xmad) [reply contracts/contracts/Tally.sol:126]: So, here is the same problem and idea as for MP.

File: contracts/contracts/Tally.sol (hunk 4 above)
Comment (0xmad) [reply contracts/contracts/Tally.sol:163]: I can try to check it, but my idea is: if we want to add more params, we don't need to change it later.

File: contracts/contracts/crypto/Verifier.sol (hunk 5 above)
Comment (0xmad) [reply contracts/contracts/crypto/Verifier.sol:37]: Yes, it works

**Comment by 0xmad** [comment]
> e2e tests work locally 🤔

I think it's something related to github because there are different flaky tests across different jobs.

File: contracts/contracts/Tally.sol (hunk 4 above)
Comment (0xmad) [reply contracts/contracts/Tally.sol:163]: So, it won't work because verifier needs dynamic size.

**Review by kittybest — APPROVED** [review]
Not sure if I understood what do these changes improve, but the code itself looks good to me

**Review by ctrlc03 — APPROVED** [review]
Great work, thanks

**Pull request merged**